pubsub/natspubsub/nats.go

```diff
@@ -0,0 +1,263 @@
+
+import (
+	"context"
+	"errors"
+	"reflect"
+
+	"github.com/nats-io/go-nats"
+	"github.com/ugorji/go/codec"
+	"gocloud.dev/gcerrors"
+	"gocloud.dev/pubsub"
+	"gocloud.dev/pubsub/driver"
+)
+
+var errNotExist = errors.New("natspubsub: topic does not exist")
+
+type topic struct {
+	nc   *nats.Conn
+	subj string
+}
+
+// For encoding we use msgpack from github.com/ugorji/go.
+// it was already imported in go-cloud and is reasonably performant.
+var mh codec.MsgpackHandle
+
+func init() {
+	// driver.Message.Metadata type
+	dm := driver.Message{}
+	mh.MapType = reflect.TypeOf(dm.Metadata)
+	mh.ExplicitRelease = true
+}
+
+// We define our own version of message here for encoding that
+// only encodes Body and Metadata. Otherwise we would have to
+// add codec decorations to driver.Message.
+type encMsg struct {
+	Body     []byte            `codec:",omitempty"`
+	Metadata map[string]string `codec:",omitempty"`
+}
+
+// OpenTopic returns a *pubsub.Topic for use with NATS.
+// We delay checking for the proper syntax here.
+func OpenTopic(nc *nats.Conn, topicName string) *pubsub.Topic {
+	return pubsub.NewTopic(openTopic(nc, topicName))
+}
+
+// openTopic returns the driver for OpenTopic. This function exists so the test
+// harness can get the driver interface implementation if it needs to.
+func openTopic(nc *nats.Conn, topicName string) driver.Topic {
+	return &topic{nc, topicName}
+}
+
+// SendBatch implements driver.Topic.SendBatch.
+func (t *topic) SendBatch(ctx context.Context, msgs []*driver.Message) error {
+	if t == nil || t.nc == nil {
+		return errNotExist
+	}
+
+	// Reuse if possible.
+	var em encMsg
+	var raw [1024]byte
+	b := raw[:0]
+	enc := codec.NewEncoderBytes(&b, &mh)
+	defer enc.Release()
+
+	for _, m := range msgs {
+		if err := ctx.Err(); err != nil {
+			return err
+		}
+		enc.ResetBytes(&b)
+		em.Body, em.Metadata = m.Body, m.Metadata
+		if err := enc.Encode(em); err != nil {
+			return err
+		}
+		if err := t.nc.Publish(t.subj, b); err != nil {
+			return err
+		}
+	}
+	// Per specification this is supposed to only return after
+	// a message has been sent. Normally NATS is very efficient
+	// at sending messages in batches on its on and also handles
+	// disconnected buffering during a reconnect event. We will
+	// let NATS handle this for now. If needed we could add a
+	// FlushWithContext() call which ensures the connected server
+	// has processed all the messages.
+	return nil
+}
+
+// IsRetryable implements driver.Topic.IsRetryable.
+func (*topic) IsRetryable(error) bool { return false }
+
+// As implements driver.Topic.As.
+func (t *topic) As(i interface{}) bool {
+	c, ok := i.(**nats.Conn)
+	if !ok {
+		return false
+	}
+	*c = t.nc
+	return true
+}
+
+// ErrorAs implements driver.Topic.ErrorAs
+func (*topic) ErrorAs(error, interface{}) bool {
+	return false
+}
+
+// ErrorCode implements driver.Topic.ErrorCode
+func (*topic) ErrorCode(error) gcerrors.ErrorCode { return gcerrors.Unknown }
```

Comment:
This should probably return gcerrors.FailedPrecondition for the error of the topic or its connection being nil.

Reply:
ok will take a look.